core/c4diagram.go

```diff
@@ -0,0 +1,355 @@
+package core
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+	"unsafe"
+
+	"github.com/kislerdm/diagramastext/core/compression"
+)
+
+type optionsPlantUMLClient struct {
+	httpClient HttpClient
+}
+
+type clientPlantUML struct {
+	options optionsPlantUMLClient
+
+	baseURL string
+}
+
+func (c *clientPlantUML) Do(v DiagramGraph) (ResponseDiagram, error) {
+	code, err := diagramGraph2plantUMLCode(v)
+	if err != nil {
+		return nil, err
+	}
+	p, err := code2Path(code)
+	if err != nil {
+		return nil, err
+	}
+
+	diagram, err := c.requestHandler("svg/" + p)
+	if err != nil {
+		return nil, err
+	}
+
+	return ResponseC4Diagram{SVG: *(*string)(unsafe.Pointer(&diagram))}, nil
+}
+
+func (c *clientPlantUML) requestHandler(p string) ([]byte, error) {
+	req, _ := http.NewRequest("GET", c.baseURL+p, nil)
+
+	resp, err := c.options.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode > 209 {
+		return nil, errors.New("error status code: " + strconv.Itoa(resp.StatusCode))
+	}
+
+	buf, err := io.ReadAll(resp.Body)
+	defer func() { _ = resp.Body.Close() }()
+	if err != nil {
+		return nil, err
+	}
+
+	return buf, nil
+}
+
+const (
+	baseURLPlanUML        = "https://www.plantuml.com/plantuml/"
+	defaultTimeoutPlanUML = 1 * time.Minute
+)
+
+// NewPlantUMLClient initiates the clientPlantUML to communicate with the plantuml server.
+func NewPlantUMLClient(optFns ...func(*optionsPlantUMLClient)) ClientGraphToDiagram {
+	o := optionsPlantUMLClient{
+		httpClient: nil,
+	}
+
+	for _, fn := range optFns {
+		fn(&o)
+	}
+
+	resolveHTTPClient(&o)
+
+	return &clientPlantUML{
+		options: o,
+		baseURL: baseURLPlanUML,
+	}
+}
+
+// WithHTTPClientPlantUML overwrite the PlantUML HTTP client.
+func WithHTTPClientPlantUML(c HttpClient) func(*optionsPlantUMLClient) {
+	return func(o *optionsPlantUMLClient) {
+		o.httpClient = c
+	}
+}
+
+func resolveHTTPClient(o *optionsPlantUMLClient) {
+	if o.httpClient == nil {
+		o.httpClient = &http.Client{Timeout: defaultTimeoutPlanUML}
+	}
+}
+
+// code2Path converts the diagram as code to the 64Bytes encoded string to query plantuml
+//
+// Example: the diagram's code
+// @startuml
+//
+//	a -> b
+//
+// @enduml
+//
+// will be converted to SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
+//
+// The resulting string to be used to generate C4 diagram
+// - as png: GET www.plantuml.com/plantuml/png/SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
+// - as svg: GET www.plantuml.com/plantuml/svg/SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
+func code2Path(s string) (string, error) {
+	zb, err := compress(*(*[]byte)(unsafe.Pointer(&s)))
+	if err != nil {
+		return "", err
+	}
+	return encode64(zb), nil
+}
+
+func compress(v []byte) ([]byte, error) {
+	var options = compression.DefaultOptions()
+	var w bytes.Buffer
+	if err := compression.Compress(&options, compression.FORMAT_DEFLATE, v, &w); err != nil {
+		return nil, err
+	}
+	return w.Bytes(), nil
+}
+
+func encode64(e []byte) string {
```

Comment:
**question** I don't remember by the top of my head how was the base64 algorithm, but this functions seems to be doing something like that, spliting into chunks of 3 bytes each and so on... The question is, do we really have to implement this encode function or is there some available function from standard packages?

Reply:
I decided not to go too deep, but rather did reverse engineering of what https://www.plantuml.com/plantuml/uml web client is doing with data before calling the endpoint to generate svg ;) Let's stick with that implementation for now and come back to it for refactoring.

Reply:
Left the "FIXME" for later refactoring.